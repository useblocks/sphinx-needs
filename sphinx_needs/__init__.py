from sphinx_needs.needs import __version__, setup  # noqa: F401
