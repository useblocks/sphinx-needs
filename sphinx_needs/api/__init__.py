# flake8: noqa

from .configuration import (
    add_dynamic_function,
    add_extra_option,
    add_need_type,
    get_need_types,
)
from .need import add_external_need, add_need, del_need, make_hashed_id
