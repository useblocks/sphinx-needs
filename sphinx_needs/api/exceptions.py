from sphinx.errors import SphinxError, SphinxWarning


class NeedsApiConfigException(SphinxError):
    """
    A configuration changes collides with the already provided configuration by the user.

    Example: An extension wants to add an already existing needs_type.
    """


class NeedsApiConfigWarning(SphinxWarning):
    pass


class NeedsNoIdException(SphinxError):
    pass


class NeedsDuplicatedId(SphinxError):
    pass


class NeedsStatusNotAllowed(SphinxError):
    pass


class NeedsTagNotAllowed(SphinxError):
    pass


class NeedsInvalidException(SphinxError):
    pass


class NeedsConstraintNotAllowed(SphinxError):
    pass


class NeedsConstraintFailed(SphinxError):
    pass


class NeedsInvalidOption(SphinxError):
    pass


class NeedsTemplateException(SphinxError):
    pass


class NeedsInvalidFilter(SphinxError):
    pass
