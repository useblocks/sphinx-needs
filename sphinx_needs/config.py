from typing import Any, Callable, Dict


class Config:
    """
    Stores sphinx-needs specific configuration values.

    This is used to avoid the usage of the sphinx internal config option, as these can be reset or cleaned in
    unspecific order during different events.

    So this Config class somehow collects possible configurations and stores it in a save way.
    """

    def __init__(self) -> None:
        self.configs: Dict[str, Any] = {}

    def add(
        self, name: str, value: Any, option_type: type = str, append: bool = False, overwrite: bool = False
    ) -> None:
        if name not in self.configs:
            self.configs[name] = option_type()
        elif not isinstance(self.configs[name], option_type):
            raise Exception(
                f"Type of needs config option {name} is {type(self.configs[name])}," f"but {option_type} is given"
            )

        if not append:
            self.configs[name] = value
        else:
            if isinstance(self.configs[name], dict):
                self.configs[name] = {**self.configs[name], **value}
            else:
                self.configs[name] += value

    def create(self, name: str, option_type: Callable[[], Any] = str, overwrite: bool = False) -> None:
        if name in self.configs and not overwrite:
            raise Exception(f"option {name} exists.")
        self.configs[name] = option_type()

    def create_or_get(self, name: str, option_type: Callable[[], Any] = str) -> Any:
        if name not in self.configs:
            self.configs[name] = option_type()
        return self.configs[name]

    def get(self, name: str) -> Any:
        return self.configs[name]


NEEDS_CONFIG = Config()
