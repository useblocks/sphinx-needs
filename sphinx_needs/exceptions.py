from __future__ import annotations

from typing import Literal

from sphinx.errors import SphinxError, SphinxWarning


class NeedsApiConfigException(SphinxError):
    """
    A configuration changes collides with the already provided configuration by the user.

    Example: An extension wants to add an already existing needs_type.
    """


class InvalidNeedException(Exception):
    """Raised when a need could not be created/added, due to a validation issue."""

    def __init__(
        self,
        type_: Literal[
            "invalid_kwargs",
            "invalid_type",
            "missing_id",
            "invalid_id",
            "duplicate_id",
            "invalid_status",
            "invalid_tags",
            "invalid_constraints",
            "invalid_jinja_content",
            "invalid_template",
            "global_option",
        ],
        message: str,
    ) -> None:
        self._type = type_
        self._message = message
        super().__init__(f"{message} [{type_}]")

    @property
    def type(
        self,
    ) -> Literal[
        "invalid_kwargs",
        "invalid_type",
        "missing_id",
        "invalid_id",
        "duplicate_id",
        "invalid_status",
        "invalid_tags",
        "invalid_constraints",
        "invalid_jinja_content",
        "invalid_template",
        "global_option",
    ]:
        return self._type

    @property
    def message(self) -> str:
        return self._message


class NeedsApiConfigWarning(SphinxWarning):
    pass


class NeedsInvalidException(SphinxError):
    pass


class NeedsConstraintFailed(SphinxError):
    pass


class NeedsInvalidFilter(SphinxError):
    pass
