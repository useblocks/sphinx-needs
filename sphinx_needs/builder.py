import os
from typing import Iterable, Optional, Set

from docutils import nodes
from sphinx import version_info
from sphinx.application import Sphinx
from sphinx.builders import Builder

from sphinx_needs.logging import get_logger
from sphinx_needs.needsfile import NeedsList
from sphinx_needs.utils import unwrap

log = get_logger(__name__)


class NeedsBuilder(Builder):
    name = "needs"
    format = "json"
    file_suffix = ".txt"
    links_suffix = None

    def write_doc(self, docname: str, doctree: nodes.document) -> None:
        pass

    def finish(self) -> None:
        env = unwrap(self.env)
        needs = env.needs_all_needs.values()  # We need a list of needs for later filter checks
        filters = env.needs_all_filters
        config = env.config
        version = getattr(config, "version", "unset")
        needs_list = NeedsList(config, self.outdir, self.srcdir)

        if config.needs_file:
            needs_file = config.needs_file
            needs_list.load_json(needs_file)
        else:
            # check if needs.json file exists in conf.py directory
            needs_json = os.path.join(self.srcdir, "needs.json")
            if os.path.exists(needs_json):
                log.info("needs.json found, but will not be used because needs_file not configured.")

        # Clean needs_list from already stored needs of the current version.
        # This is needed as needs could have been removed from documentation and if this is the case,
        # removed needs would stay in needs_list, if list gets not cleaned.
        needs_list.wipe_version(version)
        #
        from sphinx_needs.filter_common import filter_needs

        filter_string = self.app.config.needs_builder_filter
        filtered_needs = filter_needs(self.app, needs, filter_string)

        for need in filtered_needs:
            needs_list.add_need(version, need)

        for need_filter in filters.values():
            if need_filter["export_id"]:
                needs_list.add_filter(version, need_filter)

        try:
            needs_list.write_json()
        except Exception as e:
            log.error(f"Error during writing json file: {e}")
        else:
            log.info("Needs successfully exported")

    def get_outdated_docs(self) -> Iterable[str]:
        return []

    def prepare_writing(self, _docnames: Set[str]) -> None:
        pass

    def write_doc_serialized(self, _docname: str, _doctree: nodes.document) -> None:
        pass

    def cleanup(self) -> None:
        pass

    def get_target_uri(self, _docname: str, _typ: Optional[str] = None) -> str:
        return ""


def build_needs_json(app: Sphinx, _exception: Exception) -> None:

    env = unwrap(app.env)

    if not env.config.needs_build_json:
        return

    # Do not create an additional needs.json, if builder is already "needs".
    if isinstance(app.builder, NeedsBuilder):
        return

    try:
        needs_builder = NeedsBuilder(app, env)
    except TypeError:
        needs_builder = NeedsBuilder(app)
        needs_builder.set_environment(env)

    needs_builder.finish()


class NeedumlsBuilder(Builder):
    name = "needumls"

    def write_doc(self, docname: str, doctree: nodes.document) -> None:
        pass

    def finish(self) -> None:
        env = unwrap(self.env)
        needumls = env.needs_all_needumls.values()

        for needuml in needumls:
            if needuml["save"]:
                puml_content = needuml["content_calculated"]
                # check if given save path dir exists
                save_path = os.path.join(self.outdir, needuml["save"])
                save_dir = os.path.dirname(save_path)
                if not os.path.exists(save_dir):
                    os.makedirs(save_dir, exist_ok=True)

                log.info(f"Storing needuml data to file {save_path}.")
                with open(save_path, "w") as f:
                    f.write(puml_content)

    def get_outdated_docs(self) -> Iterable[str]:
        return []

    def prepare_writing(self, _docnames: Set[str]) -> None:
        pass

    def write_doc_serialized(self, _docname: str, _doctree: nodes.document) -> None:
        pass

    def cleanup(self) -> None:
        pass

    def get_target_uri(self, _docname: str, _typ: Optional[str] = None) -> str:
        return ""


def build_needumls_pumls(app: Sphinx, _exception: Exception) -> None:
    env = unwrap(app.env)

    if not env.config.needs_build_needumls:
        return

    # Do not create additional files for saved plantuml content, if builder is already "needumls".
    if isinstance(app.builder, NeedumlsBuilder):
        return

    # if other builder like html used together with config: needs_build_needumls
    if version_info[0] >= 5:
        needs_builder = NeedumlsBuilder(app, env)
        needs_builder.outdir = os.path.join(needs_builder.outdir, env.config.needs_build_needumls)
    else:
        needs_builder = NeedumlsBuilder(app)
        needs_builder.outdir = os.path.join(needs_builder.outdir, env.config.needs_build_needumls)
        needs_builder.set_environment(env)

    needs_builder.finish()
