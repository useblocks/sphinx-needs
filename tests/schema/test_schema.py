import sys
from collections.abc import Callable
from pathlib import Path
from typing import Any

import pytest
from sphinx.testing.util import SphinxTestApp
from sphinx.util.console import strip_colors

from sphinx_needs.exceptions import NeedsConfigException

CURR_DIR = Path(__file__).parent


@pytest.mark.fixture_file(
    "schema/fixtures/config.yml",
)
def test_schema_config(
    tmpdir: Path,
    content: dict[str, Any],
    make_app: Callable[[], SphinxTestApp],
    write_fixture_files: Callable[[Path, dict[str, Any]], None],
) -> None:
    # Check if test should be skipped based on min_python version
    if "mark" in content and "min_python" in content["mark"]:
        min_version = tuple(content["mark"]["min_python"])
        if sys.version_info < min_version:
            pytest.skip(
                f"Test requires Python {'.'.join(map(str, min_version))} or higher"
            )
    write_fixture_files(tmpdir, content)
    assert "exception" in content
    with pytest.raises(NeedsConfigException) as excinfo:
        make_app(srcdir=Path(tmpdir), freshenv=True)
    for snippet in content["exception"]:
        assert snippet in str(excinfo.value), (
            f"Expected exception message '{content['exception']}' not found in: {excinfo.value}"
        )


@pytest.mark.fixture_file(
    "schema/fixtures/extra_links.yml",
    "schema/fixtures/extra_options.yml",
    "schema/fixtures/network.yml",
    "schema/fixtures/unevaluated.yml",
)
def test_schemas(
    tmpdir: Path,
    content: dict[str, Any],
    make_app: Callable[[], SphinxTestApp],
    write_fixture_files: Callable[[Path, dict[str, Any]], None],
    snapshot,
) -> None:
    write_fixture_files(tmpdir, content)

    app: SphinxTestApp = make_app(srcdir=Path(tmpdir), freshenv=True)
    app.build()

    assert app.statuscode == 0
    warnings = strip_colors(app._warning.getvalue())
    assert warnings == snapshot
    app.cleanup()


@pytest.mark.parametrize(
    "test_app",
    [{"buildername": "html", "srcdir": "doc_test/doc_schema_e2e", "no_plantuml": True}],
    indirect=True,
)
def test_schema_e2e(test_app: SphinxTestApp, snapshot) -> None:
    test_app.builder.build_all()
    warnings = strip_colors(test_app._warning.getvalue())
    assert warnings == snapshot

    html = Path(test_app.outdir, "index.html").read_text()
    assert html
    # check schema validation report JSON exist
    report_json = Path(test_app.outdir, "schema_violations.json")
    assert report_json.exists(), (
        f"Expected schema validation report JSON file not found: {report_json}"
    )
