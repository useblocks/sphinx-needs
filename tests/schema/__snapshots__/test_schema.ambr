# serializer version: 1
# name: test_schema_e2e[test_app0]
  '''
  WARNING: Need 'FEAt' has validation errors:
    Severity:       violation
    Field:          id
    Need path:      FEAt
    Schema path:    [0] > local > properties > id > pattern
    User message:   id must be uppercase with numbers and underscores
    Schema message: 'FEAt' does not match '^[A-Z0-9_]+$' [sn_schema.local_fail]
  WARNING: Need 'SPEC_MISSING_APPROVAL' has validation errors:
    Severity:       violation
    Field:          
    Need path:      SPEC_MISSING_APPROVAL
    Schema path:    spec[1] > local > unevaluatedProperties
    Schema message: Unevaluated properties are not allowed ('asil', 'priority' were unexpected) [sn_schema.local_fail]
  WARNING: Need 'SPEC_MISSING_APPROVAL' has validation errors:
    Severity:       violation
    Field:          
    Need path:      SPEC_MISSING_APPROVAL
    Schema path:    spec-approved-required[2] > local > required
    User message:   Approval required due to high efforts
    Schema message: 'approved' is a required property [sn_schema.local_fail]
  WARNING: Need 'SPEC' has validation errors:
    Severity:       violation
    Field:          id
    Need path:      SPEC
    Schema path:    spec[1] > local > properties > id > pattern
    Schema message: 'SPEC' does not match '^SPEC_[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  WARNING: Need 'SPEC_SAFE_UNSAFE_FEAT' has validation errors:
    Severity:       violation
    Field:          
    Need path:      SPEC_SAFE_UNSAFE_FEAT
    Schema path:    spec[1] > local > unevaluatedProperties
    Schema message: Unevaluated properties are not allowed ('approved', 'asil', 'links', 'priority' were unexpected) [sn_schema.local_fail]
  WARNING: Need 'SPEC_SAFE_UNSAFE_FEAT' has validation errors:
    Severity:       violation
    Need path:      SPEC_SAFE_UNSAFE_FEAT > links
    Schema path:    safe-spec-[links]->safe-feat[4] > validate > network > links
    Schema message: Too few valid links of type 'links' (0 < 1) / nok: FEAT
  
      Details for FEAT
      Field:          asil
      Need path:      SPEC_SAFE_UNSAFE_FEAT > links > FEAT
      Schema path:    safe-spec-[links]->safe-feat[4] > links > local > allOf > 0 > properties > asil > enum
      Schema message: 'QM' is not one of ['A', 'B', 'C', 'D'] [sn_schema.network_contains_too_few]
  WARNING: Need 'SPEC_SAFE_ADD_UNSAFE_FEAT' has validation errors:
    Severity:       violation
    Field:          
    Need path:      SPEC_SAFE_ADD_UNSAFE_FEAT
    Schema path:    spec[1] > local > unevaluatedProperties
    Schema message: Unevaluated properties are not allowed ('approved', 'asil', 'links', 'priority' were unexpected) [sn_schema.local_fail]
  WARNING: Need 'SPEC_SAFE' has validation errors:
    Severity:       violation
    Field:          
    Need path:      SPEC_SAFE
    Schema path:    spec[1] > local > unevaluatedProperties
    Schema message: Unevaluated properties are not allowed ('approved', 'asil', 'links', 'priority' were unexpected) [sn_schema.local_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_links-beta_warn_for_extra_link]
  '''
  WARNING: Schema interface and validation are still in beta. Interface and validation logic may change when moving to a typed core implementation. [needs.beta]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_links-contains]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_links-contains_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          links
    Need path:      IMPL_1
    Schema path:    extra_links > schema > properties > links > contains
    Schema message: ['SPEC_1'] does not contain items matching the given schema [sn_schema.extra_link_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_links-inject_array]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          links
    Need path:      IMPL_1
    Schema path:    extra_links > schema > properties > links > minItems
    Schema message: ['SPEC_1'] is too short [sn_schema.extra_link_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_links-items_pattern]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_links-items_pattern_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          links.0
    Need path:      IMPL_1
    Schema path:    extra_links > schema > properties > links > items > pattern
    Schema message: 'SPEC_1' does not match '^[A-Z0-9_]{10,}' [sn_schema.extra_link_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_links-max_contains]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_links-max_contains_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          links
    Need path:      IMPL_1
    Schema path:    extra_links > schema > properties > links > contains
    Schema message: Too many items match the given schema (expected at most 1) [sn_schema.extra_link_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_links-max_items]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_links-max_items_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          links
    Need path:      IMPL_1
    Schema path:    extra_links > schema > properties > links > maxItems
    Schema message: ['SPEC_1', 'SPEC_2'] is too long [sn_schema.extra_link_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_links-min_contains]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_links-min_contains_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          links
    Need path:      IMPL_1
    Schema path:    extra_links > schema > properties > links > contains
    Schema message: Too few items match the given schema (expected at least 2 but only 1 matched) [sn_schema.extra_link_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_links-min_items]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_links-min_items_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          links
    Need path:      IMPL_1
    Schema path:    extra_links > schema > properties > links > minItems
    Schema message: ['SPEC_1'] is too short [sn_schema.extra_link_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-auto_inject_string]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-auto_inject_type]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-auto_inject_type_wrong_const]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          asil
    Need path:      IMPL_1
    Schema path:    [0] > local > properties > asil > const
    Schema message: 'QM' was expected [sn_schema.local_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-beta_warn_for_extra_option]
  '''
  WARNING: Schema interface and validation are still in beta. Interface and validation logic may change when moving to a typed core implementation. [needs.beta]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-coerce_to_boolean]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-coerce_to_boolean_from_string_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          approved
    Need path:      IMPL_1
    Schema path:    extra_options > schema
    Schema message: Field 'approved': cannot coerce 'not-a-boolean' to boolean [sn_schema.extra_option_type_error]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-coerce_to_integer]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-coerce_to_integer_from_float_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          efforts
    Need path:      IMPL_1
    Schema path:    extra_options > schema
    Schema message: Field 'efforts': cannot coerce '1.2' to integer [sn_schema.extra_option_type_error]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-coerce_to_integer_from_string_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          efforts
    Need path:      IMPL_1
    Schema path:    extra_options > schema
    Schema message: Field 'efforts': cannot coerce 'QM' to integer [sn_schema.extra_option_type_error]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-coerce_to_number]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-coerce_to_number_from_string_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          efforts
    Need path:      IMPL_1
    Schema path:    extra_options > schema
    Schema message: Field 'efforts': cannot coerce 'QM' to number [sn_schema.extra_option_type_error]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-const]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-const_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          asil
    Need path:      IMPL_1
    Schema path:    [0] > local > properties > asil > const
    Schema message: 'QM' was expected [sn_schema.local_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-default_accepted]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-enum]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-enum_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          asil
    Need path:      IMPL_1
    Schema path:    [0] > local > properties > asil > enum
    Schema message: 'E' is not one of ['QM', 'A', 'B', 'C', 'D'] [sn_schema.local_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-integer_multiple_of]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-integer_multiple_of_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          efforts
    Need path:      IMPL_1
    Schema path:    extra_options > schema > properties > efforts > multipleOf
    Schema message: 8 is not a multiple of 3 [sn_schema.extra_option_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-no_schema]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-number_multiple_of]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-number_multiple_of_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          efforts
    Need path:      IMPL_1
    Schema path:    extra_options > schema > properties > efforts > multipleOf
    Schema message: 5.0 is not a multiple of 3.3 [sn_schema.extra_option_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-required]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-required_based_on_select_field_below_threshold]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-required_based_on_select_field_missing]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-required_based_on_select_field_over_threshold]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          
    Need path:      IMPL_1
    Schema path:    [0] > local > required
    Schema message: 'asil' is a required property [sn_schema.local_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-required_missing]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          
    Need path:      IMPL_1
    Schema path:    [0] > local > required
    Schema message: 'asil' is a required property [sn_schema.local_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-set_type_string]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-string_format_date]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-string_format_date_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          start_date
    Need path:      IMPL_1
    Schema path:    extra_options > schema > properties > start_date > format
    Schema message: 'not-a-date' is not a 'date' [sn_schema.extra_option_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-string_format_date_time]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-string_format_date_time_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          start_date
    Need path:      IMPL_1
    Schema path:    extra_options > schema > properties > start_date > format
    Schema message: '2025-07-1099:99:99Z' is not a 'date-time' [sn_schema.extra_option_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-string_format_duration]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-string_format_duration_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          duration
    Need path:      IMPL_1
    Schema path:    extra_options > schema > properties > duration > format
    Schema message: 'P1Q2Q10DT2H30M' is not a 'duration' [sn_schema.extra_option_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-string_format_email]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-string_format_email_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          email
    Need path:      IMPL_1
    Schema path:    extra_options > schema > properties > email > format
    Schema message: 'not-a-mail' is not a 'email' [sn_schema.extra_option_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-string_format_time]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-string_format_time_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          time
    Need path:      IMPL_1
    Schema path:    extra_options > schema > properties > time > format
    Schema message: '26:12:13' is not a 'time' [sn_schema.extra_option_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-string_format_uri]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-string_format_uri_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          uri
    Need path:      IMPL_1
    Schema path:    extra_options > schema > properties > uri > format
    Schema message: 'examplecom' is not a 'uri' [sn_schema.extra_option_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-string_format_uuid]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-string_format_uuid_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          uuid
    Need path:      IMPL_1
    Schema path:    extra_options > schema > properties > uuid > format
    Schema message: 'deadbeef-deadbeef-deadbeef-deadbeef' is not a 'uuid' [sn_schema.extra_option_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/extra_options-string_type]
  ''
# ---
# name: test_schemas[schema/fixtures/extra_options-wrong_type]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          asil
    Need path:      IMPL_1
    Schema path:    extra_options > schema
    Schema message: Field 'asil': cannot coerce 'QM' to integer [sn_schema.extra_option_type_error]
  
  '''
# ---
# name: test_schemas[schema/fixtures/network-link_chain_hop_1_min_contains_error]
  '''
  WARNING: Need 'IMPL_SAFE' has validation errors:
    Severity:       violation
    Need path:      IMPL_SAFE > links
    Schema path:    safe-impl-[links]->safe-spec[0] > validate > network > links
    Schema message: Too few valid links of type 'links' (0 < 1) / nok: SPEC_UNSAFE
  
      Details for SPEC_UNSAFE
      Field:          asil
      Need path:      IMPL_SAFE > links > SPEC_UNSAFE
      Schema path:    safe-impl-[links]->safe-spec[0] > links > local > allOf > 0 > properties > asil > enum
      Schema message: 'QM' is not one of ['A', 'B', 'C', 'D'] [sn_schema.network_contains_too_few]
  
  '''
# ---
# name: test_schemas[schema/fixtures/network-link_chain_hop_2_min_contains_error]
  '''
  WARNING: Need 'IMPL_SAFE' has validation errors:
    Severity:       violation
    Need path:      IMPL_SAFE > links
    Schema path:    safe-impl-[links]->safe-spec-[links]->safe-req[0] > validate > network > links
    Schema message: Too few valid links of type 'links' (0 < 1) / nok: SPEC_SAFE
  
      Details for SPEC_SAFE
      Need path:      IMPL_SAFE > links > SPEC_SAFE > links
      Schema path:    safe-impl-[links]->safe-spec-[links]->safe-req[0] > links > validate > network > links
      User message:   Safe impl links to safe spec links to safe req
      Schema message: Too few valid links of type 'links' (0 < 1) / nok: REQ_UNSAFE
  
        Details for REQ_UNSAFE
        Field:          asil
        Need path:      IMPL_SAFE > links > SPEC_SAFE > links > REQ_UNSAFE
        Schema path:    safe-impl-[links]->safe-spec-[links]->safe-req[0] > links > links > local > allOf > 0 > properties > asil > enum
        Schema message: 'QM' is not one of ['A', 'B', 'C', 'D'] [sn_schema.network_contains_too_few]
  
  '''
# ---
# name: test_schemas[schema/fixtures/network-link_chain_w_refs]
  ''
# ---
# name: test_schemas[schema/fixtures/network-local_max_items]
  ''
# ---
# name: test_schemas[schema/fixtures/network-local_max_items_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          links
    Need path:      IMPL_1
    Schema path:    [0] > local > properties > links > maxItems
    Schema message: ['SPEC_1', 'SPEC_2'] is too long [sn_schema.local_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/network-local_min_items]
  ''
# ---
# name: test_schemas[schema/fixtures/network-local_min_items_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          links
    Need path:      IMPL_1
    Schema path:    [0] > local > properties > links > minItems
    Schema message: ['SPEC_1'] is too short [sn_schema.local_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/network-max_contains]
  ''
# ---
# name: test_schemas[schema/fixtures/network-max_contains_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Need path:      IMPL_1 > links
    Schema path:    [0] > validate > network > links
    Schema message: Too many valid links of type 'links' (2 > 1) / ok: SPEC_1, SPEC_2 [sn_schema.network_contains_too_many]
  
  '''
# ---
# name: test_schemas[schema/fixtures/network-max_network_levels]
  '''
  WARNING: Need 'SPEC_1' has validation errors:
    Severity:       violation
    Need path:      IMPL_1 > links > SPEC_5 > links > SPEC_4 > links > SPEC_3 > links > SPEC_2 > links > SPEC_1
    Schema path:    [0] > links > links > links > links > links
    Schema message: Maximum network validation recursion level 4 reached. [sn_schema.network_max_nest_level]
  
  '''
# ---
# name: test_schemas[schema/fixtures/network-min_contains]
  ''
# ---
# name: test_schemas[schema/fixtures/network-min_contains_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Need path:      IMPL_1 > links
    Schema path:    [0] > validate > network > links
    Schema message: Too few valid links of type 'links' (1 < 2) / ok: SPEC_1 [sn_schema.network_contains_too_few]
  
  '''
# ---
# name: test_schemas[schema/fixtures/network-schemas_in_conf]
  ''
# ---
# name: test_schemas[schema/fixtures/network-schemas_in_conf_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Need path:      IMPL_1 > links
    Schema path:    [0] > validate > network > links
    Schema message: Too few valid links of type 'links' (0 < 1) / nok: SPEC_1
  
      Details for SPEC_1
      Field:          type
      Need path:      IMPL_1 > links > SPEC_1
      Schema path:    [0] > links > local > properties > type > const
      Schema message: 'req' was expected [sn_schema.network_contains_too_few]
  
  '''
# ---
# name: test_schemas[schema/fixtures/unevaluated-unevaluated_allof_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          
    Need path:      IMPL_1
    Schema path:    [0] > local > unevaluatedProperties
    Schema message: Unevaluated properties are not allowed ('approved' was unexpected) [sn_schema.local_fail]
  
  '''
# ---
# name: test_schemas[schema/fixtures/unevaluated-unevaluated_error]
  '''
  WARNING: Need 'IMPL_1' has validation errors:
    Severity:       violation
    Field:          
    Need path:      IMPL_1
    Schema path:    [0] > local > unevaluatedProperties
    Schema message: Unevaluated properties are not allowed ('comment' was unexpected) [sn_schema.local_fail]
  
  '''
# ---
